### PR TITLE
Add a `build:ci:watch` command to watch a pipeline statuses

### LINKED ISSUE #265: Increase automated test coverage for GitLab and Pipelines
The tests for BitBucket and GitLab only run `build:project:create`.

The GitHub tests run more robust coverage, testing creating pull requests, watching for the multidev, merging the PR, etc.

The main difference is the lack of a script to watch the results of jobs on GitLab or Pipelines. The [CircleCI CLI](https://github.com/circle-cli/circle-cli) provides watch functionality out of the box.

We should have more robust tests for GitLab and BitBucket Pipelines with parity for coverage provided by the CircleCI coverage.

### PR DESCRIPTION
**Pull request by danielbachhuber**
Task list includes:

* [x] Create a new `build:ci:watch` command which supports `--branch-name=<branch>` as an argument.
* [x] Update `GitLabCIProvider` to support `getMostRecentPipelineId()` and `getPipelineStatus()`.
* [x] Update `BitbucketPipelinesProvider` to support `getMostRecentPipelineId()` and `getPipelineStatus()`.
* [x] Update `CircleCIProvider` to support `getMostRecentPipelineId()` and `getPipelineStatus()`.
* [ ] Update `CIProvider` interface to define `getMostRecentPipelineId()` and `getPipelineStatus()` methods.
* [x] Create `test-gitlab-repair.sh` and `test-bitbucket-repair.sh`; update `.circleci/config.yml` to call and watch them.
* [ ] Fetch the default branch from the Git provider instead of defaulting to the master branch.
* [x] use `terminus project:pull-request:create` from [the Terminus CLU plugin](https://github.com/pantheon-systems/terminus-clu-plugin/) instead of making direct API calls or using `hub`.

Closes #265 and #287

### CODE REVIEW

**Comment by ataylorme** [comment]
I examined the GitHub/CircleCI test coverage to determine what we need for parity between all 3 providers. The following tests should be run (in order)

- [ ] Create a new build tools project associated with a Pantheon CI bot account
  - [ ] Watch to ensure the initial CI pipeline on `master` passes
- [ ]  Set and delete a secret on the project
- [ ]  Test pull request/multidev creation
  - [ ]  Create a new Git branch
  - [ ]  Comment out `build:env:delete:pr` so it can be tested manually
  - [ ]  Commit this change above in `.ci/deploy/pantheon/dev-multidev`.
  - [ ]  Push the new branch/change to the external Git repo for the project
  - [ ]  Create a pull request based on the above branch
  - [ ]  Watch CI to ensure the pipeline related to the pull request passes
  - [ ]  Ensure a corresponding Pantheon multidev exists for the PR branch
- [ ]  Test pull request merge/multidev deletion
  - [ ]  Merge the pull request
  - [ ]  Create an additional 10 pull requests with `[ci skip]` in the title. These will not have CI run but are created to test pagination
  - [ ]  Run `build:env:delete:pr`
  - [ ]  Ensure the multidev from the original pull request has been deleted
- [ ]  Delete the test project with `buid:env:obliterate`

**Comment by ataylorme** [comment]
I've been working on GitLab. So far I have made the following commits that will need to be reverted:

- `91cd450d9437ff5933c44b2fd7cce564d58082c7` only test GitLab
  - Testing all 3 providers while only iterating on GitLab was wasteful
- `5be934f395405ae2bb1213723e4559710112ea78` Don't test specific `BUILD_TOOLS_VERSION`. For some reason `dev-265-ci-watch` was not being found by `composer create-project`, even though it exists in Packagist. This was causing the GitLab tests to fail. Since the changes here do not affect the test projects being created, only the test versions themselves, I removed the version-specific Build Tools installation. The CircleCI jobs for this PR still have the modified version
- `cf6a054d2b3a38e0e264caee65d2e31689b517ed`
  - Hard to see where GitLab pipelines failed when the projects (and job logs) were immediately deleted

Work I've completed for GitLab:

- Updated the GitLab app password and SSH keys in CircleCI with new credentials associated with a generic, Patheon IT owned account rather than one linked to Greg's user
- Created a new scripts `./.circleci/watch-ci.sh` that runs `terminus build:ci:watch`. This should be agnostic and can be used for CircleCI and BitBucket too, removing the provider specific watch commands.
- Created a `3.x` branch of the `quay.io/pantheon-public/php-ci` image that these tests use, which has an updated version of Terminus and the Terminus CLU plugin
- Updated the GitLab tests jobs to watch for pipeline success after project creation
- Started running the secret tests on GitLab

Work remaining for GitLab (see [my previous comment](https://github.com/pantheon-systems/terminus-build-tools-plugin/pull/305#issuecomment-546491614)):

- Test pull request/multidev creation
- Test pull request merge/multidev deletion

Other things I did:

- Export `TERMINUS_SITE=build-tools-$CIRCLE_BUILD_NUM` in `.circleci/set-up-globals.sh`
  - This was being set to the same value in every test script I looked at. While I've only removed it from GitLab scripts, it can likely be removed in the others as well

**Comment by ataylorme** [comment]
Another Update:

Previous tests were doing `build:project:repair` as part of creating PRs and a multidev. The issue is it was hard-coding the GitHub token into the Git remote.

I created a few new scripts to break up the testing and used SSH for the GitLab origin while cloning, so even after credential updates in `build:project:repair` there is no issue pushing and a token isn't necessary as CircleCI has an SSH key for GitLab (I still need to add/update one for BitBucket and GitHub).

The tweaks I made to the workflow are:
- Run `build:project:repair` and create new branches and PRs (to test pagination)
  - This is done in `test-pr-to-multidev.sh`
  - Only 1 branch/PR, of 13 created, actually runs pipelines. The rest are skipped with a `[ci skip]` comment
  - This still tests `build:project:repair` and the new GitLab token as `project:pull-request:create` downloads the stored token from the `dev` environment on Pantheon for use.
- The one valid merge request then has its GitLab pipeline watched
  - Using `./.circleci/watch-ci.sh test-after-repair`
- Once the PR pipeline is finished, check for multidev creation
  - `./.circleci/test-multidev-created.sh`
- Merge the PR to `master`
  - ./.circleci/test-merge-multidev-to-master.sh
- Watch the new pipeline on `master`
  - `./.circleci/watch-ci.sh master`
- Test that the PR multidev was deleted
  - ./.circleci/test-multidev-deleted-after-merge.sh

**Comment by ataylorme** [comment]
@danielbachhuber GitLab is now passing. The work remaining is to:

- Take what I've done for GitLab and replicate it for BitBucket and GitHub
  - The bash scripts can probably become even more generic. Maybe more the clone URL setup into `set-up-globals.sh` and consolidate `test-gitlab-repo.sh`, `test-github-repo.sh`, etc. into one script
  - I added an SSH key for BitBucket and a `BITBUCKET_TOKEN`envvironment variable to CircleCI so start there. I'm not sure if GitHub needs to be updated.

**Comment by ataylorme** [comment]
Another update. I was able to:

- Add `getPipelineStatus` and `getMostRecentPipelineId` methods to the BitBucket CI provider
  - [This post](https://community.atlassian.com/t5/Bitbucket-questions/I-would-like-a-way-to-know-the-last-status-of-a-pipeline-for/qaq-p/1135687#M44752) was a lot of help if working with the poorly documented BitBucket API
- Update the BitBucket tests to reflect the work I did in GitLab
  - There are still separate project create scripts that could probably be merged
- Uncomment tests for all providers
  - I also updated them to use the standard `wp` and `d8` calls on project create instead of explicitly calling `dev-master`. This will bring the tests closer to the experience users have.

Since they tests take > 30 mins to run I am going to step away so I'm not sure if BitBucket will pass

Work remaining:
- [ ] Ensure GitLab and BitBucket tests are passing
- [ ] Add `getPipelineStatus` and `getMostRecentPipelineId` methods to the Circle CI provider
- [ ] Update GitHub tests to use the updates made to GitLab and BitBucket
- [ ] Update `CIProvider` interface to define `getMostRecentPipelineId` and `getPipelineStatus` methods.

**Comment by ataylorme** [comment]
It looks like [CircleCI does **not** support getting workflow status from their API](https://discuss.circleci.com/t/trigger-workflow-through-rest-api/13931/48). They [only support getting recent builds](https://circleci.com/docs/api/#recent-builds-for-a-single-project).

However, you can get builds by branch name by adding `/tree/:branch` to the API request and the build data returned includes a workflow ID.

The only thing I could think of was looping through all builds and skipping those without the proper workflow ID.

Trying it out on GitHub + Drupal 8

**Comment by ataylorme** [comment]
It turns out CircleCI has an API version 2 and [you can get workflow status directly](https://circleci.com/docs/api/v2/#get-a-workflow).

I'm still using version `1.1` of their API to find the workflow as the v2 API doesn't have an endpoint to get workflows 😞.

**Comment by ataylorme** [comment]
I've updated the existing GitHub tests to use the new scripts/format. I _think_ once BitBucket credentials are resolved this will be good to go!

**Comment by danielbachhuber** [comment]
@ataylorme Nice work :)

**Comment by ataylorme** [comment]
@danielbachhuber can you do a code review or can we find time Monday/Tuesday to walk through this?

File: .circleci/test-bitbucket-repo.sh [hunk 1]
```diff
@@ -5,18 +5,20 @@ set -e
...
 EXTRA_ARGS="$2"
 
-BUILD_TOOLS_VERSION="dev-master"
-if [[ -n "$CIRCLE_BRANCH" ]]; then
-    BUILD_TOOLS_VERSION="dev-${CIRCLE_BRANCH}#${CIRCLE_SHA1}"
-fi
+#BUILD_TOOLS_VERSION="dev-master"
+#if [[ -n "$CIRCLE_BRANCH" ]]; then
+#    BUILD_TOOLS_VERSION="dev-${CIRCLE_BRANCH}"
+#fi
```
Comment (danielbachhuber) [inline .circleci/test-bitbucket-repo.sh]: Can this be removed if it's commented out?

File: .circleci/test-github-repo.sh [hunk 2]
```diff
@@ -5,21 +5,20 @@ set -e
...
 EXTRA_ARGS="$2"
 
-BUILD_TOOLS_VERSION="dev-master"
-if [[ -n "$CIRCLE_BRANCH" ]]; then
-    BUILD_TOOLS_VERSION="dev-${CIRCLE_BRANCH}#${CIRCLE_SHA1}"
-fi
+#BUILD_TOOLS_VERSION="dev-master"
+#if [[ -n "$CIRCLE_BRANCH" ]]; then
+#    BUILD_TOOLS_VERSION="dev-${CIRCLE_BRANCH}"
+#fi
```
Comment (danielbachhuber) [inline .circleci/test-github-repo.sh]: Can this be removed if it's commented out?

File: .circleci/test-github-repo.sh [hunk 3]
```diff
@@ -5,21 +5,20 @@ set -e
...
-TERMINUS_SITE=build-tools-$CIRCLE_BUILD_NUM
-TARGET_REPO=$GITHUB_USER/$TERMINUS_SITE
-CLONE_URL="https://github.com/${TARGET_REPO}.git"
+CLONE_URL="git@github.com:$GITHUB_USER/$TERMINUS_SITE.git"
 
 # Clear Composer cache before running build:project:create
 composer clear-cache
 
-# Build a test project on github
+# Build a test project on gitlab
```
Comment (danielbachhuber) [inline .circleci/test-github-repo.sh]: This is the `GitHub` file.

File: .circleci/test-github-repo.sh [hunk 4]
```diff
@@ -5,21 +5,20 @@ set -e
...
 terminus build:project:create -n "$SOURCE_COMPOSER_PROJECT" "$TERMINUS_SITE" --team="$TERMINUS_ORG" --email="$GIT_EMAIL" --env="BUILD_TOOLS_VERSION=$BUILD_TOOLS_VERSION" $EXTRA_ARGS
 # Confirm that the Pantheon site was created
 terminus site:info "$TERMINUS_SITE"
-# Confirm that the Github project was created
+# Confirm that the GitLab project was created
```
Comment (danielbachhuber) [inline .circleci/test-github-repo.sh]: This is the `GitHub` file.

File: .circleci/test-gitlab-repo.sh [hunk 5]
```diff
@@ -3,23 +3,21 @@
...
+EXTRA_ARGS="$2"
 
-BUILD_TOOLS_VERSION="dev-master"
-if [[ -n "$CIRCLE_BRANCH" ]]; then
-    BUILD_TOOLS_VERSION="dev-${CIRCLE_BRANCH}#${CIRCLE_SHA1}"
-fi
+#BUILD_TOOLS_VERSION="dev-master"
+#if [[ -n "$CIRCLE_BRANCH" ]]; then
+#    BUILD_TOOLS_VERSION="dev-${CIRCLE_BRANCH}"
+#fi
```
Comment (danielbachhuber) [inline .circleci/test-gitlab-repo.sh]: Can this be removed if it's commented out?

File: .circleci/test-multidev-merge-to-master.sh [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+echo -e "\nMerging branch $TEST_BRANCH_NAME into master"
+
+git checkout master
+
+git merge -m 'Merge to master' $TEST_BRANCH_NAME
+
+git push origin master
+
+# Wait 30 seconds for the merge to register and the CI pipeline to start
+sleep 30
```
Comment (danielbachhuber) [inline .circleci/test-multidev-merge-to-master.sh]: Should `sleep 30` go at the end of this file, or should it be a step on its own? I don't think it's entirely intuitive to have it at the end of this file.

File: src/Commands/CIWatchCommand.php [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        $options = [
+            'branch-name' => '',
+        ]
+    )
+    {
+        $ci_env = $this->getCIEnvironment([]);
+        // Add the environment variables from the git provider to the CI environment.
+        $ci_env->storeState('repository', $this->git_provider->getEnvironment());
+
+        // TODO Fetch the default branch from Git provider.
```
Comment (danielbachhuber) [inline src/Commands/CIWatchCommand.php]: Do we want to file a follow-up GitHub issue for this?

File: src/ServiceProviders/CIProviders/CircleCI/CircleCIProvider.php [hunk 8]
```diff
@@ -206,4 +206,136 @@ protected function remapRepositoryServiceName($serviceName)
         }
         return $serviceName;
     }
+
+    protected function apiClient()
+    {
+        $headers = [
+            // 'Content-Type' => 'application/json',
```
Comment (danielbachhuber) [inline src/ServiceProviders/CIProviders/CircleCI/CircleCIProvider.php]: Can this be removed if its commented out?

File: src/ServiceProviders/CIProviders/CircleCI/CircleCIProvider.php [hunk 9]
```diff
@@ -206,4 +206,136 @@ protected function remapRepositoryServiceName($serviceName)
...
+            'Connection' => 'keep-alive',
+            'Accept-Encoding' => 'gzip, deflate',
+            'Host' => 'circleci.com',
+            'Cache-Control' => 'no-cache',
+            'Accept' => '*/*',
+        ];
+
+        return new \GuzzleHttp\Client(
+            [
+                // 'base_uri' => 'https://circleci.com/api/v1.1',
```
Comment (danielbachhuber) [inline src/ServiceProviders/CIProviders/CircleCI/CircleCIProvider.php]: Can this be removed if it's commmented out?

File: src/ServiceProviders/CIProviders/CircleCI/CircleCIProvider.php [hunk 10]
```diff
@@ -206,4 +206,136 @@ protected function remapRepositoryServiceName($serviceName)
...
+                $errors[] = $error['message'];
+            }
+        }
+        if ($httpCode && ($httpCode >= 300)) {
+            $errors[] = "Http status code: $httpCode";
+            $message = isset($resultData['message']) ? "{$resultData['message']}." : '';
+        }
+
+        if (!empty($message) || !empty($errors)) {
+            throw new TerminusException('error: {message} {errors}', ['message' => $message, 'errors' => implode("\n", $errors)]);
```
Comment (danielbachhuber) [inline src/ServiceProviders/CIProviders/CircleCI/CircleCIProvider.php]:
```suggestion
            throw new TerminusException('CircleCI API error: {message} {errors}', ['message' => $message, 'errors' => implode("\n", $errors)]);
```

**Review by danielbachhuber — COMMENTED** [review]
Looking pretty good! I've left some comments inline with little things I saw.

I'm traveling today and tomorrow but could probably find 15 minutes to chat if there's something specific you want to discuss.

File: src/Commands/CIWatchCommand.php (hunk 7 above)
Comment (ataylorme) [reply src/Commands/CIWatchCommand.php]: I created #306

**Comment by ataylorme** [comment]
> I'm traveling today and tomorrow but could probably find 15 minutes to chat if there's something specific you want to discuss.

No need for a call then. I would just like someone else to review what is being tested since I changed how we are going about testing. I want to make sure I'm not decreasing test coverage.

One area lacking test coverage is Quicksilver pushback but since the original tests didn't cover that either I think increasing coverage is out of scope for a refactor. I opened #307 so that work can be done in the future.

**Comment by danielbachhuber** [comment]
> I would just like someone else to review what is being tested since I changed how we are going about testing. I want to make sure I'm not decreasing test coverage.

Ok. It looks pretty good to me but I'm not 100% confident.

As a sanity check, maybe build a table of what's being tested before and after?

**Comment by ataylorme** [comment]
> As a sanity check, maybe build a table of what's being tested before and after?

I did that on paper already. The only major change (aside from script names/job ordering) is the coverage around `build:project:repair`.

That command is being run before pull requests are created with `project:pull-request:create`, which fetches the token stored in the build meta data and validates authentication so I'm confident the credential cycling in `build:project:repair` is still covered by the change.

**Comment by ataylorme** [comment]
It seems the GitLab API is returning `503` and `502` errors. I'll wait a bit then re-run the tests

**Comment by ataylorme** [comment]
@greg-1-anderson I'm assigning you to handle updating the BitBucket app password in CircleCI once it comes back from Pantheon IT, making sure the tests all pass, and merging this PR.

**Comment by ataylorme** [comment]
Woo! Got [BitBucket (and the rest of the tests) passing](https://circleci.com/workflow-run/84479fbc-7686-464c-8e85-864f1ab49479)

**Pull request closed without merging**